File: golang/445. 两数相加 II.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	st1 := make([]int, 0)
	st2 := make([]int, 0)
	for l1 != nil {
		st1 = append(st1, l1.Val)
		l1 = l1.Next
	}
	for l2 != nil {
		st2 = append(st2, l2.Val)
		l2 = l2.Next
	}
	var ans *ListNode
	carry := 0
	for len(st1) != 0 || len(st2) != 0 || carry != 0 {
		a, b := 0, 0
		if len(st1) != 0 {
			a = st1[len(st1)-1]
			st1 = st1[:len(st1)-1]
		}
		if len(st2) != 0 {
			b = st2[len(st2)-1]
			st2 = st2[:len(st2)-1]
		}
		t := a + b + carry
		carry = t / 10
		curnode := &ListNode{Val: t % 10, Next: ans}
		ans = curnode
	}
	return ans
}

File: 刷题/cpp/445. 两数相加 II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* l3=nullptr;
        stack<int> st1, st2;
        while(l1!=nullptr){
            st1.push(l1->val);
            l1=l1->next;
        }
        while(l2!=nullptr){
            st2.push(l2->val);
            l2=l2->next;
        }
        int carry=0;
        while(!st1.empty()||!st2.empty()||carry!=0){
            int a=st1.empty()?0:st1.top();
            int b=st2.empty()?0:st2.top();
            if(!st1.empty()) st1.pop();
            if(!st2.empty()) st2.pop();
            int cur=a+b+carry;
            carry=cur/10;
            cur%=10;
            auto curnode=new ListNode(cur);
            curnode->next=l3;
            l3=curnode;
        }
        return l3;
    }   
};